pkg/initializers/dataset/cache.py

```diff
@@ -0,0 +1,295 @@
+                                    "ports": [{"containerPort": 50051}],
+                                }
+                            ],
+                        },
+                    },
+                    "workerTemplate": {
+                        "spec": {
+                            "serviceAccountName": f"{train_job_name}-sa",
+                            "containers": [
+                                {
+                                    "name": "worker",
+                                    "image": cache_image,
+                                    "command": ["worker"],
+                                    "args": ["0.0.0.0", "50051"],
+                                    "resources": {
+                                        "limits": {
+                                            "cpu": worker_cpu,
+                                            "memory": worker_mem,
+                                        },
+                                        "requests": {
+                                            "cpu": worker_cpu,
+                                            "memory": worker_mem,
+                                        },
+                                    },
+                                    "env": env_vars,
+                                    "ports": [{"containerPort": 50051}],
+                                }
+                            ],
+                        }
+                    },
+                },
+            },
+        }
+
+        # Create LeaderWorkerSet
+        custom_api.create_namespaced_custom_object(
+            group="leaderworkerset.x-k8s.io",
+            version="v1",
+            namespace=namespace,
+            plural="leaderworkersets",
+            body=lws_body,
+        )
+        logging.info(f"Created LeaderWorkerSet {train_job_name}-cache")
+
+        # Create Service
+        service = client.V1Service(
+            metadata=client.V1ObjectMeta(
+                name=f"{train_job_name}-cache-service",
+                namespace=namespace,
+                owner_references=[owner_ref],
+            ),
+            spec=client.V1ServiceSpec(
+                selector={"app": f"{train_job_name}-cache-head"},
+                ports=[
+                    client.V1ServicePort(protocol="TCP", port=50051, target_port=50051)
+                ],
+            ),
+        )
+
+        try:
+            core_v1.create_namespaced_service(namespace=namespace, body=service)
+            logging.info(f"Created Service {train_job_name}-cache-service")
+        except ApiException as e:
+            if e.status == 409:
+                logging.info(
+                    f"Service {train_job_name}-cache-service already exists, skipping creation"
+                )
+            else:
+                raise
+
+        # Wait for LeaderWorkerSet to become ready
+        lws_name = f"{train_job_name}-cache"
+
+        while True:
+            try:
```

Comment:
It would be good to refactor this with `watch()` API, e.g. https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/backends/kubernetes/backend.py#L349
Please can you add the TODO statement ?